libraries/AP_Compass/AP_Compass.cpp

```diff
@@ -502,20 +518,27 @@ void Compass::_detect_backends(void)
         return;
     }
 
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4
+    if (AP_BoardConfig::get_board_type() == AP_BoardConfig::PX4_BOARD_PIXHAWK2) {
+        // default to disabling LIS3MDL on pixhawk2 due to hardware issue
+        _driver_type_mask.set_default(1U<<DRIVER_LIS3MDL);
+    }
+#endif
+    
 /*
   macro to add a backend with check for too many backends or compass
   instances. We don't try to start more than the maximum allowed
  */
-#define ADD_BACKEND(backend, name, external) \
-    do { _add_backend(backend, name, external);     \
+#define ADD_BACKEND(driver_type, backend, name, external)   \
+    do { if (_driver_enabled(driver_type)) { _add_backend(backend, name, external); } \
        if (_backend_count == COMPASS_MAX_BACKEND || \
```

Comment:
Unrelated to this PR and I think I've written this before somewhere, but this macro doesn't work. *_add_backend* will panic if this condition is true, so it will never reach here.